File: sorting/mergeSort/merge.py
def merge(l1,l2):
    n1,n2,m1,m2=len(l1),len(l2),0,0
    l=[]
    while(m1+m2<n1+n2):
        if(m1==n1):
            l.append(l2[m2])
            m2+=1
        elif(m2==n2):
            l.append(l1[m1])
            m1+=1
        elif(l1[m1]<l2[m2]):
            l.append(l1[m1])
            m1+=1
        elif(l2[m2]<=l1[m1]):
            l.append(l2[m2])
            m2+=1
    return l
def sort(l):
    n=len(l)
    if(n<=1):
        return l
    else:
        return merge(sort(l[0:n//2]),sort(l[n//2:n]))
l=[5,3,1,2,4]
l=sort(l)
print(l)

File: sorting/mergeSort/merge.c
#include<stdio.h>
void merge(int *arr,int low,int mid,int high){
    int n1=mid-low+1;
    int n2=high-mid;
    int l[n1];
    int r[n2];
    for(int i=0;i<n1;i++){
        l[i]=arr[low+i];
    }
    for(int i=0;i<n2;i++){
        r[i]=arr[mid+i+1];
    }
    int i=0,j=0,k=low;
    while(i<n1&&j<n2){
        if(l[i]<=r[j]){
            arr[k++]=l[i++];
        }
        else if(l[i]>r[j]){
            arr[k++]=r[j++];
        }
    }
    while(i<n1){
        arr[k++]=l[i++];
    }
    while(j<n2){
        arr[k++]=r[j++];
    }
}
void mergesort(int *arr,int low,int high){
    if(low<high){
        int mid=(low+high)/2;
        mergesort(arr,low,mid);
        mergesort(arr,mid+1,high);
        merge(arr,low,mid,high);
    }
}
int main(){
    int arr[5]={5,4,3,2,1};
    mergesort(arr,0,4);
    for(int i=0;i<5;i++){
        printf("%d ",arr[i]);
    }
}